java/arcs/core/data/Plan.kt

```diff
@@ -21,8 +21,33 @@ import arcs.core.util.lens
 open class Plan(
     // TODO(cromwellian): add more fields as needed (e.g. RecipeName, etc for debugging)
     val particles: List<Particle>,
-    val arcId: String? = null
+    val annotations: List<Annotation>
 ) {
+    // TODO(b/155796088): get rid of these ctors and use annotations in PlanPartition and Recipe.
+    constructor(particles: List<Particle>, arcId: String? = null) :
+        this(
+            particles,
+            if (arcId != null)
+                listOf(Annotation("arcId", mapOf("id" to AnnotationParam.Str(arcId))))
+            else emptyList()
+        )
+
+    constructor(particles: List<Particle>, annotations: List<Annotation>, arcId: String) :
+        this(particles, annotations) {
+        require(this.arcId == arcId)
+    }
+
+    val arcId: String?
+        get() {
+            return annotations.find { it.name == "arcId" }?.let {
+                val idParam = requireNotNull(it.params["id"])
+                return when (idParam) {
+                    is AnnotationParam.Str -> idParam.value
+                    else -> throw UnsupportedOperationException("Unexpected arcId param $idParam")
```

Comment:
This error message should probably say it's the wrong type.

Also a more appropriate exception type would be IllegalStateException

Reply:
done